### PR TITLE
Add custom localization with one-camera-at-a-time pose estimation

### LINKED ISSUE #34: Update Photonvision to use single-cam PhotonPoseEstimator
Some form of custom localization system will need to be implemented to deal with the multiple camera poses

### PR DESCRIPTION
**Pull request by kchenturtles**
Closes #34

### CODE REVIEW

File: src/main/java/frc/robot/Constants.java [hunk 1]
```diff
@@ -52,6 +54,7 @@ public static final class PathConstants {
     public static final double KP = 0.00000094597;
     public static final double KI = 0;
     public static final double KD = 0;
+    public static final double MAX_VISION_TO_ODOMETRY_ERROR = 1;
```
Comment (aidan-mundy) [inline src/main/java/frc/robot/Constants.java]: What does this mean? Is there a unit for this?

File: src/main/java/frc/robot/Constants.java [hunk 2]
```diff
@@ -89,9 +92,18 @@ public static final class AutonConstants {
   }
 
   public static final class VisionConstants {
-    public static final Transform3d FRONT_LEFT_CAMERA_TO_ROBOT = new Transform3d();
-    public static final Transform3d FRONT_RIGHT_CAMERA_TO_ROBOT = new Transform3d();
-    public static final Transform3d REAR_CAMERA_TO_ROBOT = new Transform3d();
+    public static final Transform3d ROBOT_TO_FRONT_LEFT_CAMERA =
```
Comment (aidan-mundy) [inline src/main/java/frc/robot/Constants.java:94]: Something seems wrong with these positions

File: src/main/java/frc/robot/Constants.java [hunk 3]
```diff
@@ -89,9 +92,18 @@ public static final class AutonConstants {
...
+            new Translation3d(Units.inchesToMeters(9.766), Units.inchesToMeters(7.852), Units.inchesToMeters(28.242)),
+            new Rotation3d());
+    public static final Transform3d ROBOT_TO_FRONT_RIGHT_CAMERA =
+        new Transform3d(
+            new Translation3d(Units.inchesToMeters(9.766), -Units.inchesToMeters(7.852), Units.inchesToMeters(28.242)),
+            new Rotation3d());
+    public static final Transform3d ROBOT_TO_REAR_CAMERA =
+        new Transform3d(new Translation3d(0, -Units.inchesToMeters(12.1585), Units.inchesToMeters(20.749)),
+            new Rotation3d(0, 0, Math.PI));
+    public static final double MAINTAIN_CAMERA_THRESHOLD = 0.5;
```
Comment (aidan-mundy) [inline src/main/java/frc/robot/Constants.java]: What does this mean? Is there a unit for this?

If this is the confidence threshold, it should probably be higher than 50%

File: src/main/java/frc/robot/Constants.java (hunk 1 above)
Comment (kchenturtles) [reply src/main/java/frc/robot/Constants.java]: whoops, don't need that anymore

File: src/main/java/frc/robot/Constants.java (hunk 2 above)
Comment (kchenturtles) [reply src/main/java/frc/robot/Constants.java:94]: more specific?

File: src/main/java/frc/robot/Constants.java (hunk 3 above)
Comment (kchenturtles) [reply src/main/java/frc/robot/Constants.java]: yeah it's confidence threshold

File: src/main/java/frc/robot/Constants.java (hunk 2 above)
Comment (citrus-melon) [reply src/main/java/frc/robot/Constants.java:94]: Aidan showed me that the x distance between the front and back cameras is shorter than it should be.

Also, the magic number for the back camera is 0.07 m, but I thought it was more like 0.2 m?

**Comment by citrus-melon** [comment]
Suggested edit:
```diff
diff --git a/src/main/java/frc/robot/subsystems/DriveBase.java b/src/main/java/frc/robot/subsystems/DriveBase.java
index a0ccf16..4ec9055 100644
--- a/src/main/java/frc/robot/subsystems/DriveBase.java
+++ b/src/main/java/frc/robot/subsystems/DriveBase.java
@@ -311,14 +311,12 @@ public class DriveBase extends SubsystemBase {
    */
   public void periodic() {
     poseEstimator.update(getHeading(), getLeftEncoderDistanceMeters(), getRightEncoderDistanceMeters());
-    try {
       EstimatedRobotPose estimatedVisionPose = this.photonVision.getCurrentPose();
-      SmartDashboard.putString("PhotonVision Pose", estimatedVisionPose.estimatedPose.toString());
-      poseEstimator.addVisionMeasurement(estimatedVisionPose.estimatedPose,
-          estimatedVisionPose.timestampSeconds);
-    } catch (NullPointerException npe) {
-
-    }
+      if (estimatedVisionPose != null) {
+        SmartDashboard.putString("PhotonVision Pose", estimatedVisionPose.estimatedPose.toString());
+        poseEstimator.addVisionMeasurement(estimatedVisionPose.estimatedPose,
+            estimatedVisionPose.timestampSeconds);
+      }
 
     SmartDashboard.putNumber("Gyro", this.getHeading().getDegrees());
     SmartDashboard.putString("Pose", this.getPose().toString());

```

File: src/main/java/frc/robot/subsystems/DriveBase.java [hunk 4]
```diff
@@ -318,12 +312,10 @@ public void tankDrive(double leftSpeed, double rightSpeed, boolean squaredInputs
...
-      Optional<Pair<Pose3d, Double>> visionEstimatedPose = photonVision.getEstimatedGlobalPose();
-      if (visionEstimatedPose.isPresent()) {
-        SmartDashboard.putString("PhotonVision Pose", visionEstimatedPose.get().getFirst().toPose2d().toString());
-        poseEstimator.addVisionMeasurement(visionEstimatedPose.get().getFirst().toPose2d(),
-            visionEstimatedPose.get().getSecond());
-      }
+      EstimatedRobotPose estimatedVisionPose = this.photonVision.getCurrentPose();
+      SmartDashboard.putString("PhotonVision Pose", estimatedVisionPose.estimatedPose.toString());
+      poseEstimator.addVisionMeasurement(estimatedVisionPose.estimatedPose,
+          estimatedVisionPose.timestampSeconds);
```
Comment (citrus-melon) [inline src/main/java/frc/robot/subsystems/DriveBase.java:318]:
```suggestion
      EstimatedRobotPose estimatedVisionPose = this.photonVision.getCurrentPose();
      if (estimatedVisionPose != null) {
          SmartDashboard.putString("PhotonVision Pose", estimatedVisionPose.estimatedPose.toString());
          poseEstimator.addVisionMeasurement(estimatedVisionPose.estimatedPose,
          estimatedVisionPose.timestampSeconds);
      }
```

**Review by citrus-melon — CHANGES_REQUESTED** [review]

**Comment by citrus-melon** [comment]
Approving blindly

**Pull request closed without merging**